sdk/python/kfp/components/component_factory.py

```diff
@@ -58,26 +58,41 @@ def _python_function_name_to_component_name(name):
     return name_with_spaces[0].upper() + name_with_spaces[1:]
 
 
-_INSTALL_PYTHON_PACKAGES_SCRIPT = '''
+def _make_index_url_options(pip_index_urls: Optional[List[str]]) -> str:
+    if not pip_index_urls:
+        return ""
+
+    index_url = pip_index_urls[0]
+    extra_index_urls = pip_index_urls[1:]
+
+    options = [f"--index-url {index_url} --trusted-host {index_url} "]
+    options.extend(
+        f"--extra-index-url {extra_index_url} --trusted-host {extra_index_url} "
+        for extra_index_url in extra_index_urls)
+
+    return " ".join(options)
+
+
+def _get_packages_to_install_command(
+        package_list: Optional[List[str]] = None,
+        pip_index_urls: Optional[List[str]] = None) -> List[str]:
+
+    if not package_list:
+        return []
+
+    concat_package_list = ' '.join(
+        [repr(str(package)) for package in package_list])
+    index_url_options = _make_index_url_options(pip_index_urls)
+
+    install_python_packages_script = f'''
```

Comment:
nit: I think we can still define this string template outside the function, right? so the indentation doesn't look weird here--on a first glance, I thought this was due to some bad merge.

Reply:
Alternatively, use `textwrap.dedent`

e.g.: https://github.com/kubeflow/pipelines/blob/5dcc6d9910372eee0b37cf31723910ed5f6fe98e/sdk/python/kfp/components/pipeline_task_test.py#L63-L76

Reply:
Done